### ISSUE TITLE
convergence bug?

### ISSUE THREAD

**Issue by hugoguh** [pavanramkumar/pyglmnet#65]
I noticed it while working on cvpyglmnet:

![screen shot 2016-05-09 at 2 01 10 pm](https://cloud.githubusercontent.com/assets/3664942/15124619/857f8570-15ee-11e6-84fc-f33565b78db3.png)

If you notice there is a weird behavior on the training performance: the training set deviance is supposed to always go down as `reg_lambda` approaches zero (or log(Lambda) becomes more negative).

Here’s some code for you to replicate it:
<sup> doesn’t happen always, for instance try using `np.random.seed(0)` </sup>
```python
import numpy as np
import scipy.sparse as sps
from sklearn.preprocessing import StandardScaler
import scikits.bootstrap as boot
from pyglmnet import GLM
np.random.seed(42)

# create an instance of the GLM class
reg_lambda = np.exp(np.linspace(-10, -3, num=100))
model = GLM(distr='poisson', verbose=False, alpha=0.95, reg_lambda=reg_lambda)

n_samples, n_features = 10000, 100

# coefficients
beta0 = np.random.normal(0.0, 1.0, 1)
beta = sps.rand(n_features, 1, 0.1)
beta = np.array(beta.todense())

# training data
Xr = np.random.normal(0.0, 1.0, [n_samples, n_features])
yr = model.simulate(beta0, beta, Xr)

# testing data
Xt = np.random.normal(0.0, 1.0, [n_samples, n_features])
yt = model.simulate(beta0, beta, Xt)

# fit Generalized Linear Model
scaler = StandardScaler().fit(Xr)
```
fit and compute deviance
<sup> using average deviance, just a scaler diference but better measure when cross-validating because different folds might have different number of elements, use the commented line for deviance </sup>

```python
yrhat = model.fit_predict(scaler.transform(Xr), yr)
#dev_t = [model.deviance(yr, i) for i in yr_t_hat]
dev_t = [model.deviance(yr, i)/float(np.shape(yrhat)[1]) for i in yrhat]
```
now plot

```python
%matplotlib inline
import matplotlib.pyplot as plt
upto = 60
plt.plot(np.log(model.reg_lambda[0:upto]), dev_t[0:upto], '-o', c='k')
plt.xlabel('log(Lambda)')
plt.ylabel('Poisson Deviance')
```
Here’s the output:
![screen shot 2016-05-09 at 1 48 14 pm](https://cloud.githubusercontent.com/assets/3664942/15124213/b7ba5e5e-15ec-11e6-9aaa-2d6066ac04d3.png)

play around with the range of `reg_lambdas` and also try with other `random.seed` (different simulated `Xr and `yr`). 

Any idea on where it is coming from? At first I thought it was a warm start effect @pavanramkumar  says it starts by fitting the larger `reg_lambdas`.
This might give a hint, it seems to not depend on the actual value of `reg_lambda`: when it happens it seems to always happen towards the end, see if I use a slightly different range of `reg_lambdas` when instantiating the model:
```python
reg_lambda = np.exp(np.linspace(-8, -3, num=100))
```
![screen shot 2016-05-09 at 2 03 17 pm](https://cloud.githubusercontent.com/assets/3664942/15124758/1979111a-15ef-11e6-8959-f7d38ed12b1e.png)

**Comment by jasmainak** [comment]
The last plot just looks like a zoomed version of the first plot to me. Did you try other metrics like RMSE or so?

**Comment by pavanramkumar** [comment]
renaming a few issue titles for clarity

**Comment by jasmainak** [comment]
I think this is related to https://github.com/glm-tools/pyglmnet/pull/226. You just need to use a smaller `learning_rate`

**Comment by jasmainak** [comment]
@pavanramkumar @hugoguh this seems to be fixed now with latest code. Just run this:

```py
import numpy as np
import scipy.sparse as sps
from sklearn.preprocessing import StandardScaler
import scikits.bootstrap as boot
from pyglmnet import GLM, simulate_glm
np.random.seed(42)

# create an instance of the GLM class
reg_lambda = np.exp(np.linspace(-10, -3, num=100))

n_samples, n_features = 10000, 100

# coefficients
beta0 = np.random.normal(0.0, 1.0, 1)[0]
beta = sps.rand(n_features, 1, 0.1)
beta = np.array(beta.todense())[:, 0]

# training data
Xr = np.random.normal(0.0, 1.0, [n_samples, n_features])
yr = simulate_glm('poisson', beta0, beta, Xr)

# testing data
Xt = np.random.normal(0.0, 1.0, [n_samples, n_features])
yt = simulate_glm('poisson', beta0, beta, Xt)

# fit Generalized Linear Model
scaler = StandardScaler().fit(Xr)

dev_t = list()
for rl in reg_lambda:
    glm = GLM(distr='poisson', verbose=True, alpha=0.95, reg_lambda=rl)
    glm.fit(scaler.transform(Xr), yr)
    dev_t.append(glm.score(scaler.transform(Xr), yr))


import matplotlib.pyplot as plt
upto = 60
plt.plot(np.log(reg_lambda[0:upto]), dev_t[0:upto], '-o', c='k')
plt.xlabel('log(Lambda)')
plt.ylabel('Poisson Deviance')
```

I get:

![Screen Shot 2020-08-14 at 4 58 26 PM](https://user-images.githubusercontent.com/15852194/90292177-6feff180-de4f-11ea-9d87-1ec84337d812.png)


can you close if you agree?
